### DISCUSSION TITLE
p:button outcome evaluated early

### DISCUSSION THREAD

**Discussion by lumpynose** [primefaces/community#260] [PrimeFaces] [answered]
contactlistparam.xhtml contains an h:datatable with

```                
                <h:column>
                    <f:facet name="header">
                    View/Delete
                    </f:facet>
                    <p:link value="View/Delete"
                        outcome="/param/contactviewparam.xhtml?faces-redirect=true">
                        <f:param name="paramId" value="#{contact.id}" />
                    </p:link>
                </h:column>
```
contactviewparam.xhtml contains

```
                <p:button value="Delete"
                    outcome="#{contactViewParam.deleteContact(contactViewParam.contact.id)}">
                </p:button>
```

ContactViewParam.java contains

```
    public String deleteContact(final Long id) {
        this.log.debug("deleteContact(id): paramId: {}", this.paramId);

        this.contactService.deleteContact(id);

        return "/param/contactlistparam.xhtml?faces-redirect=true";
    }
```

I've also tried it with

```
    public String deleteContact() {
        this.log.debug("deleteContact(): paramId: {}", this.paramId);

        final Long id = Long.valueOf(this.paramId);

        this.contactService.deleteContact(id);

        return "/param/contactlistparam.xhtml?faces-redirect=true";
    }
```

When I click the View/Delete link in contactlistparam.xhtml it sends me to contactviewparam.xhtml but it also runs  ContactViewParam's deleteContact method.  When I click the button it sends me to contactlistparam.xhtml as expected but doesn't run the deleteContact method (again).  What am I doing wrong for it to run the deleteContact method before I click the button?

```
        <dependency>
            <groupId>jakarta.platform</groupId>
            <artifactId>jakarta.jakartaee-api</artifactId>
            <version>10.0.0</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.primefaces</groupId>
            <artifactId>primefaces</artifactId>
            <version>12.0.0</version>
            <classifier>jakarta</classifier>
        </dependency>
```

**Comment by tandraschko** [comment, marked as answer]
IMO you miss-use outcome; just use the action attribute on a commandButton.

**Reply by lumpynose** [reply]
Thanks, that fixed it when I changed it in contactviewparam.xhtml.  contactlistparam.xhtml is unchanged.

I don't understand though; can you explain why using p:link didn't work?

Edit: I think I understand now after thinking some more; with commandButton and commandLink the action= expression isn't evaluated until you click on it, while with link and button the outcome= is evaluated sooner (or something).

**Reply by tandraschko** [reply]
Check the docs, Outcome is something complete different and evaluated at different time then a action
